kernel/src/engine/arrow_utils.rs

```diff
@@ -1074,6 +1073,31 @@ fn parse_json_impl(json_strings: &StringArray, schema: ArrowSchemaRef) -> DeltaR
     Ok(concat_batches(&schema, output.iter())?)
 }
 
+pub(crate) fn filter_to_record_batch(
+    filtered_data: FilteredEngineData,
+) -> DeltaResult<RecordBatch> {
+    // Honor the new contract: if selection vector is shorter than the number of rows,
+    // then all rows not covered by the selection vector are assumed to be selected
+    let (underlying_data, mut selection_vector) = filtered_data.into_parts();
+    let batch = extract_record_batch(&*underlying_data)?;
+    let num_rows = batch.num_rows();
+
+    let result_batch = if selection_vector.is_empty() {
+        // If selection vector is empty, write all rows per contract.
+        batch.clone()
+    } else {
+        // Extend the selection vector with `true` for uncovered rows
+        if selection_vector.len() < num_rows {
+            selection_vector.resize(num_rows, true);
+        }
```

Comment:
Technically, it's safe to `resize` unconditionally. There's already an internal length check to nop it in case the length is already correct. And it's safe to truncate an oversized selection vector.
```suggestion
        selection_vector.resize(num_rows, true);
```